sklearn/ensemble/_forest.py

```diff
@@ -398,13 +408,63 @@ def fit(self, X, y, sample_weight=None):
         if self.oob_score:
             self._set_oob_score(X, y)
 
+        if self.feature_importances_type == "permutation":
+            self._set_oob_permutation_importance(X, y)
+
         # Decapsulate classes_ attributes
         if hasattr(self, "classes_") and self.n_outputs_ == 1:
             self.n_classes_ = self.n_classes_[0]
             self.classes_ = self.classes_[0]
 
         return self
 
+    def _set_oob_permutation_importance(self, X, y):
+        """
+        Calculate out of bag predictions and score."""
+        X = check_array(X, dtype=DTYPE, accept_sparse=False)
+
+        n_samples = y.shape[0]
+
+        n_predictions = np.zeros(n_samples)
+
+        n_samples_bootstrap = _get_n_samples_bootstrap(
+            n_samples, self.max_samples
+        )
+
+        all_imp = np.zeros((self.n_estimators, X.shape[1]))
+        for i, estimator in enumerate(self.estimators_):
+            random_state = check_random_state(estimator.random_state)
+
+            unsampled_indices = _generate_unsampled_indices(
+                estimator.random_state, n_samples, n_samples_bootstrap
+            )
+
+            scorer = check_scoring(estimator, scoring=self.scoring)
+            baseline = scorer(
+                estimator, X[unsampled_indices, :], y[unsampled_indices, :]
+            )
+
+            for col in range(X.shape[1]):
+                X_permuted = X.copy()
+                random_state.shuffle(X_permuted[:, col])
+                curr_perf = scorer(
+                    estimator,
+                    X_permuted[unsampled_indices, :],
+                    y[unsampled_indices, :],
+                )
+                all_imp[i, col] = baseline - curr_perf
+
+            n_predictions[unsampled_indices] += 1
+
+        self._oob_permutation_importance = all_imp.mean(axis=0)
```

Comment:
I could imagine that we could store a `Bunch` object with the mean, std and all importance as in `permutation_importance`:

```
    result: Bunch
        Dictionary-like object, with the following attributes.
        importances_mean: ndarray, shape (n_features, )
            Mean of feature importance over n_repeats.
        importances_std: ndarray, shape (n_features, )
            Standard deviation over n_repeats.
        importances: ndarray, shape (n_features, n_repeats)
            Raw permutation importance scores.
```